File: leetcode-cn/2.两数相加.py
#
# @lc app=leetcode.cn id=2 lang=python3
#
# [2] 两数相加
#


# @lc code=start
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        sign = 0
        ans = ListNode()
        p = ans
        while l1 is not None and l2 is not None:
            sum_ = l1.val + l2.val

            p.next = ListNode((sum_ + sign) % 10)
            sign = (sum_ + sign) // 10

            l1 = l1.next
            l2 = l2.next
            p = p.next

        while l1 is not None:
            sum_ = l1.val

            p.next = ListNode((sum_ + sign) % 10)
            sign = (sum_ + sign) // 10

            l1 = l1.next
            p = p.next

        while l2 is not None:
            sum_ = l2.val

            p.next = ListNode((sum_ + sign) % 10)
            sign = (sum_ + sign) // 10

            l2 = l2.next
            p = p.next

        if sign != 0:
            p.next = ListNode(sign)

        return ans.next


# @lc code=end

File: leetcode-cn/C++/2.两数相加.cpp
/*
 * @lc app=leetcode.cn id=2 lang=cpp
 *
 * [2] 两数相加
 */

// @lc code=start
// struct ListNode {
//     int val;
//     ListNode* next;
//     ListNode()
//         : val(0)
//         , next(nullptr)
//     {
//     }
//     ListNode(int x)
//         : val(x)
//         , next(nullptr)
//     {
//     }
//     ListNode(int x, ListNode* next)
//         : val(x)
//         , next(next)
//     {
//     }
// };
#include <iostream>
using namespace std;
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2)
    {
        ListNode* new_head = new ListNode();
        ListNode* cur = new_head;
        int carry = 0;

        while (l1 != nullptr or l2 != nullptr or carry != 0) {
            int a = 0;
            int b = 0;

            if (l1) {
                a = l1->val;
                l1 = l1->next;
            }
            if (l2) {
                b = l2->val;
                l2 = l2->next;
            }

            int res = a + b + carry;
            carry = res / 10;
            res = res % 10;

            ListNode* node = new ListNode(res);
            cur->next = node;
            cur = cur->next;
        }

        return new_head->next;
    }
};
// @lc code=end
